src/ports/postgres/modules/stats/pred_metrics.py_in

```diff
@@ -0,0 +1,562 @@
+
+
+def metric_agg_help_msg(schema_madlib, message, agg_name, **kwargs):
+
+    if not message:
+        help_string = """
+------------------------------------------------------------
+                        SUMMARY
+------------------------------------------------------------
+Functionality: Evaluate prediction results using metric functions.
+
+This module provides a set of prediction accuracy metrics. It is a support
+module for several machine learning algorithms that require metrics to validate
+their models. The function will take "prediction" and "observation" values to
+calculate the desired metric. Grouping is supported by all of these functions.
+    """
+    elif message.lower().strip() in ['usage', 'help', '?']:
+        help_string = """
+------------------------------------------------------------
+                        USAGE
+------------------------------------------------------------
+SELECT {schema_madlib}.{agg_name}(
+    'table_in',     -- Name of the input table
+    'table_out',    -- Table name to store the metric results
+    'pred_col',     -- Column name containing prediction results
+    'obs_col',      -- Column name containing observed (actual) values
+    'grouping_cols' -- Comma-separated list of columns to use as group-by
+);
+    """
+    else:
+        help_string = "No such option. Use {schema_madlib}.{agg_name}('usage')"
+    return help_string.format(**locals())
+
+
+def _get_r2_score_sql(table_in, pred_col, obs_col, grp_col_str=None):
+    """ Generate the SQL query to compute r2 score.
+
+    This function abstracts the SQL to calculate r2 score from actually building
+    the output table. This allows reusing the query for adjusted r2 function.
+
+    Args:
+        @param table_in: str, Input table name containing the data
+        @param pred_col: str, Column name containing the predictions
+        @param obs_col: str, Column name containing the actual observed class
+        @param grp_col_str: str, Comma-separated list of columns to group by
+
+    Definition:
+        r2 = 1 - SS_res / SS_tot
+        where SS_res = sum (pred - obs)^2
+              SS_tot = sum (obs - mean)^2
+
+    """
+    if grp_col_str:
+        grp_out_str = grp_col_str + ","
+        grp_by_str = "GROUP BY " + grp_col_str
+        partition_str = "PARTITION BY " + grp_col_str
+    else:
+        grp_out_str = grp_by_str = partition_str = ""
+    return """
+            SELECT {grp_out_str}
+                1 - ssres/sstot AS r2_score
+            FROM (
+                SELECT {grp_out_str}
+                       sum(({obs_col} - mean)^2) as sstot,
+                       sum(({pred_col} - {obs_col})^2) AS ssres
+                FROM(
+                    SELECT {grp_out_str}
+                           {pred_col}, {obs_col},
+                           avg({obs_col}) OVER ({partition_str}) as mean
+                    FROM {table_in}
+                ) x {grp_by_str}
+            ) y
```

Comment:
Why not 
```sql
SELECT 
     {grp_out_str}
     1 - avg(({pred_col} - {obs_col})^2)/var_pop({obs_col}) AS r2_score
FROM {table_in} {grp_by_str}
```
It is simpler, faster (2-3 times in my quick experiments) and numerically stable (avoid large sum)